src/main/scala/com/fulcrumgenomics/util/AssignPrimers.scala

```diff
@@ -86,17 +93,46 @@ class AssignPrimers
       mateAmpliconIdentifierTag = mateAmpliconIdentifierTag
     )
 
-    reader.foreach { rec =>
-      val recAmplicon = detector.findPrimer(rec=rec)
-      labeller.label(
-        rec = rec,
-        recAmplicon = recAmplicon,
-        mateAmplicon = if (rec.unpaired) None else detector.findMatePrimer(rec=rec)
-      )
-      writer.write(rec)
-      progress.record(rec)
+    if (annotateAll) {
+      // All alignments will be assigned a primer independently, and then grouped by read end. An primer assignment will
+      // be made on a given read end only if there is one and only one unique primer assignment.
+      val sorter = Bams.sorter(SamOrder.Coordinate, reader.header)
```

Comment:
I think this is a problem on a couple of fronts.  1) Since the `writer` is created using the `reader.header` above it's possible that the SO in the header and the actual ordering of the reads won't match.  2) I think the expectation is probably to produce output in the same order as the input since that's what the tool generally does.

I'd prefer to move the declaration of `writer` inside the if/else then on the `if` branch we could extract the `SamOrder` from the input header and use that to drive the `sorter` and perhaps even skip re-sorting if the input file was already query sorted or grouped?